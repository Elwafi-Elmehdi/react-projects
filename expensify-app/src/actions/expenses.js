import { v4 as uuidv4 } from "uuid";

// Actions for Expensify

// ADD_EXPENSE
export const addExpense = ({
	title = "no title",
	note = "",
	amount = 0.0,
	createdAt = 0,
} = {}) => ({
	type: "ADD_EXPENSE",
	expense: {
		id: uuidv4(),
		title,
		note,
		amount,
		createdAt,
	},
});

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
	type: "REMOVE_EXPENSE",
	id,
});
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
	type: "EDIT_EXPENSE",
	id,
	updates,
});
