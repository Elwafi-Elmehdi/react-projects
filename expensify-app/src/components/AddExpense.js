import React from "react";
const AddExpensePage = () => <div>this is from my Add Expense cmp</div>;
export default AddExpensePage;
