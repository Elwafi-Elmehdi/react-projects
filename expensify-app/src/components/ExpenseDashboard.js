import React from "react";

const ExpenseDashboardPage = () => <div>this is from my dashboard cmp</div>;

export default ExpenseDashboardPage;
