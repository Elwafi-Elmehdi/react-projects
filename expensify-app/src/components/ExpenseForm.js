import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import "react-dates/lib/css/_datepicker.css";

export default class ExpenseForm extends React.Component {
	state = {
		description: "",
		amount: 0,
		note: "",
		createdAt: moment(),
		calenderFocused: false,
		error: "",
	};
	onSubmit = (e) => {
		e.preventDefault();
		if (!this.state.amount || !this.state.description) {
			this.setState(() => ({ error: "Please entre a valid expense" }));
		} else {
			this.setState(() => ({ error: "" }));
			this.props.onSubmit({
				title: this.state.description,
				note: this.state.note,
				amount: parseFloat(this.state.amount),
				createdAt: this.state.createdAt.valueOf(),
			});
		}
	};
	onFocusChange = ({ focused }) => {
		this.setState(() => ({ calenderFocused: focused }));
	};
	onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({ createdAt }));
		}
	};
	onChangeDescription = (e) => {
		const description = e.target.value;
		this.setState((state) => ({
			description,
		}));
	};
	onChangeNote = (e) => {
		const note = e.target.value;
		this.setState((state) => ({
			note,
		}));
	};
	onChangeAmount = (e) => {
		const amount = e.target.value;
		if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
			this.setState(() => ({ amount }));
		}
	};
	render() {
		return (
			<div>
				{this.state.error && <p>{this.state.error}</p>}
				<form onSubmit={this.onSubmit}>
					<input
						type="text"
						value={this.state.description}
						placeholder="Title"
						autoFocus
						onChange={this.onChangeDescription}
					/>
					<input
						type="text"
						value={this.state.amount}
						placeholder="Amount"
						onChange={this.onChangeAmount}
					/>
					<SingleDatePicker
						date={this.state.createdAt}
						onDateChange={this.onDateChange}
						focused={this.state.calenderFocused}
						onFocusChange={this.onFocusChange}
						numberOfMonths={1}
						isOutsideRange={() => false}
					/>
					<textarea
						placeholder="Note"
						value={this.state.note}
						onChange={this.onChangeNote}
					></textarea>

					<button>Add Expense</button>
				</form>
			</div>
		);
	}
}
