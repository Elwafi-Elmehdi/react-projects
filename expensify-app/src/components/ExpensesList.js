import React from "react";
import { connect } from "react-redux";

const ExpensesList = (props) => (
	<div>
		<p>Hello from Expenses list</p>
	</div>
);

export default connect()(ExpensesList)

export default ExpensesList;
