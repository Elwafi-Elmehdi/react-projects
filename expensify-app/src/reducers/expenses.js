// Expense Reducer

const expenseInitState = [];

export default (state = expenseInitState, action) => {
	// Never mutate the state , always return a new state
	switch (action.type) {
		case "ADD_EXPENSE":
			return [...state, action.expense];
		case "REMOVE_EXPENSE":
			return state.filter(({ id }) => id !== action.id);
		case "EDIT_EXPENSE":
			return state.map((expense) => {
				if (action.id === expense.id) {
					return {
						...expense,
						...action.updates,
					};
				} else {
					return expense;
				}
			});
		default:
			return state;
	}
};
