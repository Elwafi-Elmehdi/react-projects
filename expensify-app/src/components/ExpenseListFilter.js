import React from "react";
import { connect } from "react-redux";
import { addTextFilter, sortByAmount, sortByDate } from "../actions/filters";

const ExpenseListFilter = (props) => (
	<div>
		<input
			type="text"
			value={props.filters.text}
			onChange={(e) => {
				props.dispatch(addTextFilter(e.target.value));
			}}
		/>
		<select
			value={props.filters.sortBy}
			onChange={(e) => {
				if (e.target.value === "date") {
					props.dispatch(sortByDate());
				} else if (e.target.value === "amout") {
					props.dispatch(sortByAmount());
				}
			}}
		>
			<option value="date">Date</option>
			<option value="amount">Amount</option>
		</select>
	</div>
);

const mapStateToProps = (state) => ({
	filters: state.filters,
});

export default connect(mapStateToProps)(ExpenseListFilter);
