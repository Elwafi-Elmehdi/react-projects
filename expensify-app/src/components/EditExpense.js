import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { editExpense, removeExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {

}
const EditExpensePage = (props) => {
	return (
		<div>
			<ExpenseForm
				onSubmit={(expense) => {
					props.dispatch(editExpense(props.match.params.id, expense));
					props.history.push("/");
				}}
				expense={props.expenses.find(
					(expense) => expense.id === props.match.params.id
				)}
			/>
			<button
				onClick={() => {
					props.dispatch(
						removeExpense({ id: props.match.params.id })
					);
					props.history.push("/");
				}}
			>
				Remove
			</button>
		</div>
	);
};
const mapStateToProps = (state) => ({ expenses: state.expenses });
const mapDispatchToProps = (dispatch) => {
	return {

	}
}
export default connect(mapStateToProps,mapDispatchToProps())(EditExpensePage);
