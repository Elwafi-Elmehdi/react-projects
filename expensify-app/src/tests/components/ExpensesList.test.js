import React from "react";
import { ExpensesList } from "../../components/ExpensesList";
import { shallow } from "enzyme";
import { expenses } from "../selectors/expenses.test";

test("Should render expense list with expenses passed as props", () => {
	const wrapper = shallow(<ExpensesList expenses={expenses} />);
	expect(wrapper).toMatchSnapshot();
});
