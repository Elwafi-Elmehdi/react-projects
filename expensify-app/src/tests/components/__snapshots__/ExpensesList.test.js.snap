// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render expense list with expenses passed as props 1`] = `
<div>
  <ExpensesListItem
    expense={
      Object {
        "amount": 45,
        "createdAt": 78542,
        "id": 1,
        "note": "Testing is awesome",
        "title": "Hegoeddwad",
      }
    }
    key="1"
  />
  <ExpensesListItem
    expense={
      Object {
        "amount": 3000,
        "createdAt": -345600000,
        "id": 2,
        "note": "",
        "title": "Car Payment",
      }
    }
    key="2"
  />
  <ExpensesListItem
    expense={
      Object {
        "amount": 700,
        "createdAt": 604800000,
        "id": 3,
        "note": "",
        "title": "Real-state taxes",
      }
    }
    key="3"
  />
</div>
`;

exports[`Should render expense list without props 1`] = `
<div>
  <p>
    No Expenses found.
  </p>
</div>
`;
